### PR TITLE
Removed tomcat::config::context::valve::resource_name

### PR DESCRIPTION
**Pull request by h-haaks**
Attribute/Property 'name' is not valid for Context/Valve.
Fixes MODULES-10855

### CODE REVIEW

**Comment by h-haaks** [comment]
I don't really know why my last commit break the travis run...

I had a green run in https://travis-ci.org/github/puppetlabs/puppetlabs-tomcat/builds/740343420
Then I added an acceptance test and I can't get acceptande tests to pass.

It seems to be a problem between facter-4.0.44, puppet-6.19.1, bolt-2.24.1 and Logging::Logger

`
.rake aborted!
NoMethodError: undefined method `trace' for #<Logging::Logger:0x0000000005ee3c70>
Did you mean?  trap
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/bolt-2.24.1/lib/bolt/pal/logging.rb:15:in `handle'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/log.rb:196:in `block in newmessage'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/log.rb:195:in `each'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/log.rb:195:in `newmessage'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/log.rb:315:in `initialize'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/log.rb:88:in `new'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/log.rb:88:in `create'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/util/logging.rb:277:in `block in setup_facter_logging!'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/facter-4.0.44/lib/facter/framework/logging/logger.rb:64:in `debug'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/facter-4.0.44/lib/facter/framework/core/fact_loaders/fact_loader.rb:33:in `load_internal_facts'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/facter-4.0.44/lib/facter/framework/core/fact_loaders/fact_loader.rb:24:in `load'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/facter-4.0.44/lib/facter/framework/core/fact_manager.rb:20:in `resolve_facts'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/facter-4.0.44/lib/facter.rb:468:in `resolve_fact'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/facter-4.0.44/lib/facter.rb:333:in `value'
/home/travis/build/puppetlabs/puppetlabs-tomcat/vendor/bundle/ruby/2.5.0/gems/puppet-6.19.1/lib/puppet/settings.rb:77:in `hostname_fact'
`

**Comment by h-haaks** [comment]
I see that puppet-community-rangefinder reports a breaking change..
Should I leave the  resource_name parameter in the type and just issue a deprecation warning if it is used?

**Comment by h-haaks** [comment]
To run tests locally I'm using pdk 1.18.1.0 and 'pdk bundle install'
which gives

Installing facter 4.0.44
Using puppet 6.17.0
Installing bolt 2.25.0

**Comment by h-haaks** [comment]
Seems like the green build in https://travis-ci.org/github/puppetlabs/puppetlabs-tomcat/builds/740343420 is using
Installing facter 4.0.44
Installing puppet 6.19.0
Installing bolt 2.32.0

While the failing build https://travis-ci.org/github/puppetlabs/puppetlabs-tomcat/jobs/740478889 is using
Using facter 4.0.44
Installing puppet 6.19.1
Installing bolt 2.24.1

**Comment by david22swan** [comment]
@h-haaks Apologies for the wait on responding to this. The issue you're seeing is a known one and one that I am currently looking at. Sorry for any inconvenience it's caused you and I hope to get a solve in place quickly.

**Comment by h-haaks** [comment]
Thanks @david22swan, I was not expecting any response during the weekend :)
While you are working on fixing that issue, do you have any input on 

> I see that puppet-community-rangefinder reports a breaking change..
> Should I leave the resource_name parameter in the type and just issue a deprecation warning if it is used?

**Comment by david22swan** [comment]
@h-haaks Just closing and re-opening to kick a clean travis run.
In regards to your change I am unsure if it is needed. After trying the manifest shown in your ticket I was able to get a successful run on a test machine, only requiring the updating of my 'catalina_base' to match my machine.
It seems to me that either there is something else causing the issue, this seem's like it may be unlikely, or the issue is specific to the Tomcat you are using.
In which case I don't think I can go ahead and merge the PR as it is.
Do you have any more information that you can provide, i.e. what Tomcat you are using, or anything you deem relevant?

**Comment by h-haaks** [comment]
@david22swan I have tested this on RHEL7.8, adoptopenjdk-8-hotspot with the following releases of tomcat:
apache-tomcat-8.5.55.tar.gz
apache-tomcat-8.5.56.tar.gz
apache-tomcat-8.5.59.tar.gz

There is no waring when i use 8.5.55, but starting with 8.5.56 I'm getting the waring
`WARNING [localhost-startStop-1] org.apache.tomcat.util.digester.SetPropertiesRule.begin [SetPropertiesRule]{Context/Valve} Setting property 'name' to 'RemoteIpValve' did not find a matching property.`

If you look at def documentation
https://tomcat.apache.org/tomcat-7.0-doc/config/valve.html
https://tomcat.apache.org/tomcat-8.5-doc/config/valve.html
https://tomcat.apache.org/tomcat-9.0-doc/config/valve.html
There is not a single valve class with a name attibute.

File: manifests/config/context/valve.pp [hunk 1]
```diff
@@ -19,8 +16,7 @@
 #
 define tomcat::config::context::valve (
   Enum['present','absent'] $ensure = 'present',
-  $resource_name                   = $name,
```
Comment (david22swan) [inline manifests/config/context/valve.pp:22]: Why was the name removed?

**Comment by h-haaks** [comment]
@david22swan I read trought my logs again when using apache-tomcat-8.5.55.tar.gz, and found my waring there as well.
I tweeked some of the accentance tests locally so that I could exec into the container to view cataline.out, but I could not see anything.
Then I changed spec_helper_acceptance_local.rb to not lock down tomcat8 to 8.5.39
Still nothing in the logs..

Finally I created a new acceptance spec file to create a setup equivalent to my usage of the module.
Running tests using this spec file on main branch with tomcat8 lockdown removed, 
[fix_valve_spec.zip](https://github.com/puppetlabs/puppetlabs-tomcat/files/5480589/fix_valve_spec.zip)
produces the warning I'm seeing.

I don't want to commit the spec file, because I don't know the impact on other tests, but I'll attach it here so that you can have a look.

**Comment by h-haaks** [comment]
@david22swan I ran my spec without changes in spec_helper_acceptance_local.rb and I get the same warning on tomcat 8.5.39!

**Comment by david22swan** [comment]
@h-haaks 
I have been unable to reproduce your reported error from the information that you have given me. I have brought in a colleague and they also have not been able to reproduce it.
Looking into the context element in more detail I have tracked down some documentation that states that the name variable is supported and so should be included, see here:
https://tomcat.apache.org/tomcat-8.5-doc/config/context.html#Context_Parameters
I am sorry that I have not been more help but I cannot in good conscience merge this change without being able to reproduce the original problem.

**Comment by h-haaks** [comment]
@david22swan the link you posted is for <Context><Paramerter/></Context> which is managed by tomcat::config::context::parameter and not tomcat::config::context::valve.

Would it be better if I turn it a bit around and include the resource_name param again as optional and not as the namevar?

**Comment by h-haaks** [comment]
As I understand the tomcat documentation a valve could be added to conf/server.xml as 
`<Server><Service name=..><Engine name=..><Valve className=../></Engine></Service></Server>` or
`<Server><Service name=..><Enginename=..><Host name=..><Valve className=../></Host></Engine></Service></Server>`
or context.xml as
`<Context><Valve className=../></Context>`
The <Valve> element should have the same elements where ever you deside to put it.

Why then is there not any code in tomcat::config::server::valve to set the name attibute when it manages
`<Server><Service name=..><Enginename=..><Host name=..><Valve className=../></Host></Engine></Service></Server>`

**Comment by david22swan** [comment]
@h-haaks Apologies, I had linked the wrong part of the page. If you read through the introduction section of the linked document you will see that it states that each context object requires its own unique name to act as an identifier within the host.
https://tomcat.apache.org/tomcat-8.5-doc/config/context.html#Introduction

**Comment by h-haaks** [comment]
@david22swan if you look at https://tomcat.apache.org/tomcat-8.5-doc/config/context.html#Access_Logs they don't even have a name attribute in the samples ...
For a Valve the className attribute is the qnique name, and I guess thats why it's bold in the docs for each valve type
https://tomcat.apache.org/tomcat-8.5-doc/config/valve.html#Access_Log_Valve/Attributes

Please verify this with the apache tomcat community before you close this issue.

**Comment by h-haaks** [comment]
Here is the simplest example I could think of to show you.
Just docker build and run this Dockerfile and you'll have to see the warning!

Just remove the `name="AccessLogValve"` attribute, build and run again and the warning is gone :)
[Dockerfile.zip](https://github.com/puppetlabs/puppetlabs-tomcat/files/5483969/Dockerfile.zip)

**Comment by h-haaks** [comment]
If there should be a custom valve implementation out there with a name property it could always added by puting it in $additional_attributes

**Comment by h-haaks** [comment]
Once more, I'll be more than happy to rewrite the class so that it is not breaking backwards compatability and log warings to user.
When doing this I would also like to add a $class_name parameter to allign naming.
Then I'll deprecate both resource_name and resource_type.

This will also allign with tomcat::config::server::valve

**Comment by david22swan** [comment]
@h-haaks 
Apologies but after further discussion with my team at large we have made the collective decision not to merge this PR as it will compromise the ability to set multiple valve object's within the context.xml. 
As an example, without it Puppet would be unable to disambiguate between valves with the same className, but different attributes, as shown below:
```puppet
# https://tomcat.apache.org/tomcat-8.5-doc/config/context.html#Access_Logs
# <Valve name="access_log"
#        className="org.apache.catalina.valves.AccessLogValve"
#        prefix="localhost_access_log"
#        suffix=".txt"
#        pattern="common"/>
# <Valve name="access_log_rare"
#        className="org.apache.catalina.valves.AccessLogValve"
#        prefix="localhost_access_log_rare"
#        suffix=".txt"
#        pattern="rare"/>
tomcat::config::context::valve {
  'access_log':
    ensure                => present,
    resource_type         => 'org.apache.catalina.valves.AccessLogValve',
    additional_attributes => {
      prefix  => 'localhost_access_log',
      suffix  => '.txt',
      pattern =>'common'
    };
  'access_log_rare':
    ensure                => present,
    resource_type         => 'org.apache.catalina.valves.AccessLogValve',
    additional_attributes => {
      prefix  => 'localhost_access_log_rare',
      suffix  => '.txt',
      pattern =>'rare'
    };
}
```
Due to this we need to keep the code as it is and accept the warnings. I am sorry that I was unclear in my previous responses, my understanding of this module and it's types is not the best.
If you can, as you had offered, rewrite this pr to be backwards compatible and make the name attribute optional then this may change however.

**Comment by h-haaks** [comment]
Ah, now I understand why the name attribute is needed :) I did not realize that a single valve class could be used multiple times in the same context.

Given this I still think the resource_type parameter should be the 'namevar', and resource_name made optional.
With this and a good description for the resource_name parameter explaining why it is necessary in some cases, and that using it will generate a tomcat warning when applications are deployed.

I also think the same changes should be implemented in tomcat::config::server::valve.

Futher more I think it would be the best if parameter names in tomcat::config::server::valve and tomcat::config::server::valve where made equal.

If you don't mind I would like to give it a try.

**Comment by david22swan** [comment]
@h-haaks 
Thanks a lot for the response and really Looking forward to reviewing the changes
My only concern is the swapping of `resource_type` to the namevar, I'm not sure that this would work as the namevar is the part of the manifest that Puppet uses as an identifier and so must be unique.
As a thought, an idea could be to default the namevar to be the `resource_type` only if a `resource_type` is not given separately, and have it act as `resource_name` if the `resource_type` is provided separately. This could allow you to gain the functionality that you wish while still maintaining what is there.

**Comment by h-haaks** [comment]
@david22swan 
Since a valve can't exist without a className I was actually planning to use an augeas search with both attributtes if resource_name is given.

**Comment by h-haaks** [comment]
I'll rename my branch and create a new PR on my changes, so I'll close this one now.
Thanks again @david22swan for your patience :)

**Pull request closed without merging**